xarray/tests/test_backends.py

```diff
@@ -2665,6 +2666,86 @@ def create_store(self):
             yield group
 
 
+@requires_zarr
+class TestZarrWriteEmpty(TestZarrDirectoryStore):
+    @contextlib.contextmanager
+    def temp_dir(self):
+        with tempfile.TemporaryDirectory() as d:
+            store = os.path.join(d, "test.zarr")
+            yield d, store
+
+    @contextlib.contextmanager
+    def roundtrip_dir(
+        self,
+        data,
+        store,
+        save_kwargs=None,
+        open_kwargs=None,
+        allow_cleanup_failure=False,
+    ):
+        if save_kwargs is None:
+            save_kwargs = {}
+        if open_kwargs is None:
+            open_kwargs = {}
+
+        data.to_zarr(store, **save_kwargs, **self.version_kwargs)
+        with xr.open_dataset(
+            store, engine="zarr", **open_kwargs, **self.version_kwargs
+        ) as ds:
+            yield ds
+
+    @pytest.mark.parametrize("write_empty", [True, False])
+    def test_write_empty(self, write_empty):
+        if not write_empty:
+            expected = ["0.1.0", "1.1.0"]
+        else:
+            expected = [
+                "0.0.0",
+                "0.0.1",
+                "0.1.0",
+                "0.1.1",
+                "1.0.0",
+                "1.0.1",
+                "1.1.0",
+                "1.1.1",
+            ]
+
+        ds = xr.Dataset(
+            data_vars={
+                "test": (
+                    ("Z", "Y", "X"),
+                    np.array([np.nan, np.nan, 1.0, np.nan]).reshape((1, 2, 2)),
+                )
+            }
+        )
+
+        if has_dask:
+            ds["test"] = ds["test"].chunk((1, 1, 1))
+            encoding = None
+        else:
+            encoding = {"test": {"chunks": (1, 1, 1)}}
+
+        with self.temp_dir() as (d, store):
+            ds.to_zarr(
+                store,
+                mode="w",
+                encoding=encoding,
+                write_empty_chunks=write_empty,
```

Comment:
Does it make sense to allow this to be controlled on a per-variable basis? If so, `encoding` would be the right place to specify it. I don't have an opinion, just a question

Reply:
It's an option for Zarr array creation so it would originally be specified on a per-variable basis; however, the issue is that there is no way to specify it when appending data to an existing store, as there is an error raised when `encoding` is not empty.

Reply:
What this reveals is that our usage of `encoding` is overloaded. There are evidently two distinct types of information that we can put in `encoding`:
- Specifications for how to store  the data on disk, e.g. dtype, compression, etc, that must be consistent for all subsequent writes to the Zarr array
- Runtime choices, like `write_empty_chunks`, that can be different for each write (and that don't necessarily need to be known to read the data back)

Perhaps not necessary for this PR, but perhaps we could consider explicitly distinguishing these two distinct types of encoding. Maybe one solution is to eventually remove `write_empty_chunks` and all such runtime choices from `encoding` and have them only be supported as kwargs.

Reply:
@rabernat That sounds like it would be a good idea. I'm sure there are probably more cases where this lack of distinction could cause a problem.